kratos/testing/distributed_test_case.cpp

```diff
@@ -52,26 +51,75 @@ bool DistributedTestCase::IsDisabled() const
     return !IsEnabled();
 }
 
+bool DistributedTestCase::IsDistributedTest() const
+{
+    return true;
+}
+
 std::string DistributedTestCase::Info() const
 {
     return "Distributed test case " + Name();
 }
 
 void DistributedTestCase::CheckRemoteFailure()
 {
-    bool success_on_this_rank = GetResult().IsSucceed();
+    // The data communicator
     const DataCommunicator& r_comm = ParallelEnvironment::GetDefaultDataCommunicator();
-    bool global_success = r_comm.AndReduceAll(success_on_this_rank);
-    if (success_on_this_rank && !global_success)
-    {
-        TestCaseResult remote_failure(GetResult());
-        remote_failure.SetToFailed();
-        remote_failure.SetErrorMessage("Test was reported as successful on this rank, but failed on a different rank.");
-        SetResult(remote_failure);
+
+    // First we check if tests are run on all ranks
+    const bool run_on_this_rank = GetResult().IsRun();
+    const bool global_run = r_comm.AndReduceAll(run_on_this_rank);         // It is run on all ranks
+    if (!global_run) {
+        if (run_on_this_rank) {
+            TestCaseResult remote_run(GetResult());
+            remote_run.Reset();
+            remote_run.SetErrorMessage("Test was reported as run on this rank, but did not run on a different rank.");
+            SetResult(remote_run);
+        } else {
+            TestCaseResult rank_result(GetResult());
+            rank_result.Reset();
+            rank_result.SetErrorMessage("Test was not run on this rank.");
+            SetResult(rank_result);
+        }
+
+        return;
+    }
+
+    // Test skipped on at least one rank.
+    const bool skipped_on_this_rank = GetResult().IsSkipped();
+    const bool global_skipped = r_comm.OrReduceAll(skipped_on_this_rank);  // It is skipped on some rank
+    if (global_skipped) {
+        if (!skipped_on_this_rank) { // This rank skipped, but another rank skipped.
+            TestCaseResult remote_skip(GetResult());
+            remote_skip.SetToSkipped();
+            SetResult(remote_skip);
+        } else { // Test failed at least on this rank
+            TestCaseResult rank_result(GetResult());
+            rank_result.SetToSkipped();
+            SetResult(rank_result);
+        }
+
+        return;
+    }
+
+    // Test failed on at least one rank.
+    const bool success_on_this_rank = GetResult().IsSucceed();
+    const bool global_success = r_comm.AndReduceAll(success_on_this_rank); // It succeeds on all ranks
```

Comment:
I don't like the fact that we are doing three separate communication steps for each individual test. Would it be practical to do a single communication with a vector? I remember that I decided not to implement communication for std::vector<bool> due to its particularities, but there should be a version that converts it to a vector of ints internally for the communication.